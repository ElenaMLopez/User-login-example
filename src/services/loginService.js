function hasToken(token_key) {
  if (localStorage.getItem(token_key)) {
    return true;
  } else {
    return false;
  }
}

const loginService = (to, from, next) => {
  if (to.path != "/login" && hasToken("karumi_user_token")) {
    next();
  } else {
    next("login");
  }
};

export default loginService;
