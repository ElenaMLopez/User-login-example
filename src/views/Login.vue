<template>
  <div class="login-container">
    <img alt="Karumi logo" src="../assets/logo.png" />
    <LoginForm />
  </div>
</template>
<script>
import LoginForm from "@/components/LoginForm.vue";

export default {
  name: "Login",
  components: {
    LoginForm
  }
};
</script>
<style lang="scss" scoped>
.login-container {
  border-radius: 5px;
  border: 1.5px solid #cccccc;
  margin: 100px auto;
  padding: 100px 40px 20px;
  width: 280px;
}
</style>
