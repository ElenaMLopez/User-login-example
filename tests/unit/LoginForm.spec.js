require("./mock-local-storage");

import { shallowMount, mount, createLocalVue } from "@vue/test-utils";
import VueRouter from "vue-router";
import flushPromises from "flush-promises";
// Investigate why this library has error and how to fix router testing
// import { toHaveRouteName } from 'vue-test-utils-helpers';
import mockRouter from "./__mocks__/mockRouter";
import Login from "@/components/LoginForm.vue";
import loginService from "@/services/loginService";

var axios = require("axios");
var MockAdapter = require("axios-mock-adapter");

const localVue = createLocalVue();
localVue.use(VueRouter);
const router = new VueRouter();
const mockUser = {
  token: "fake_token"
};

describe("Login component from Home view", () => {
  it("Should render component", () => {
    const wrapper = shallowMount(Login);
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("LoginForm.vue", () => {
    it("Shoud have a username input", () => {
      const wrapper = shallowMount(Login);
      const userInput = wrapper.find("#userEmail");

      expect(userInput.exists()).toBe(true);
    });
    it("Shoud have a password input", () => {
      const wrapper = shallowMount(Login);
      const userInput = wrapper.find("#userPassword");

      expect(userInput.exists()).toBe(true);
    });
    it("Should have a login button", () => {
      const wrapper = shallowMount(Login);
      const logoutButton = wrapper.find(".primarybtn");
      expect(logoutButton.exists()).toBe(true);
    });
  });

  describe("Test click on login button", () => {
    it("onclick", async () => {
      const wrapper = shallowMount(Login);
      const _sendData = jest.fn();
      const clickButton = wrapper.find(".primarybtn");

      await wrapper.trigger("click", { method: _sendData, router });
      expect(clickButton.exists()).toBeTruthy();
    });

    it("Should navigate to Home on successful login", async () => {
      loginService.isLoginValid = jest.fn().mockResolvedValue(true);
      const localVue = createLocalVue();

      localVue.use(VueRouter);
      const router = mockRouter.mock();

      const wrapper = shallowMount(Login, {
        localVue,
        router
      });

      wrapper.find(".primarybtn").trigger("click");
      await flushPromises();

      // TO DO: Investigate why this dont works.
      // expect(wrapper).toHaveRouteName('home');
    });
  });
  describe("Save token at localStorage", () => {
    it("return ok and token to localstorage", async () => {
      const $router = { push: jest.fn() };
      const wrapper = shallowMount(Login);
      var mock = new MockAdapter(axios);

      wrapper.vm.$router = $router;
      mock.onGet("/API/user.json").reply(200, [mockUser]);

      localStorage.setItem("karumi_user_token", mockUser.token);
      await wrapper.find(".primarybtn").trigger("click");

      expect(localStorage.getItem("karumi_user_token")).toEqual(mockUser.token);
    });
  });
});
