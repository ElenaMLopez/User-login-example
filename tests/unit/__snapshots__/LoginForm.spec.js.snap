// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component from Home view Should render component 1`] = `
<div>
  <form type="submit" class="login-form"><input id="userEmail" type="email" name="email" placeholder="User Email" data-cy="input-userEmail" class="login-form__user-input"> <input id="userPassword" type="password" name="password" placeholder="Password" data-cy="input-userPass" class="login-form__user-input"> <button data-cy="send-form" class="form-button primarybtn">
      Login
    </button>
    <!---->
  </form>
</div>
`;
